release/releaser_test.go

```diff
@@ -527,6 +532,154 @@ func Test_UpdateList(t *testing.T) {
+		manifests: mockManifests,
+		repo:      checkout,
+		registry:  mockRegistry,
+	}
+	spec := update.ContainerSpecs{
+		ContainerSpecs: map[flux.ResourceID][]update.ContainerUpdate{
+			hwSvcID: { // success
+				{
+					Container: helloContainer,
+					Current:   oldRef,
+					Target:    newHwRef,
+				},
+			},
+			testSvcID: { // fail
+				{
+					Container: testContainer,
+					Current:   oldRef,
+					Target:    oldRef,
+				},
+			},
+		},
+		Kind: update.ReleaseKindExecute,
+	}
+
+	_, err := Release(ctx, spec, log.NewNopLogger())
+	assert.Error(t, err)
+
+	spec.IgnoreFailedControllers = true
+	results, err := Release(ctx, spec, log.NewNopLogger())
+	assert.NoError(t, err)
+
+	expected := update.Result{
+		hwSvcID: update.ControllerResult{
+			Status: update.ReleaseStatusSuccess,
+			PerContainer: []update.ContainerUpdate{{
+				Container: helloContainer,
+				Current:   oldRef,
+				Target:    newHwRef,
+			}},
+		},
+		testSvcID: update.ControllerResult{
+			Status: update.ReleaseStatusFailed,
+			Error:  fmt.Sprintf(update.ContainerTagMismatch, testContainer),
+		},
+	}
+	assert.Equal(t, expected[hwSvcID], results[hwSvcID])
+	assert.Equal(t, expected[testSvcID], results[testSvcID])
+}
+
+func Test_UpdateContainers(t *testing.T) {
+	cluster := mockCluster(hwSvc, lockedSvc)
+	checkout, cleanup := setup(t)
+	defer cleanup()
+	ctx := &ReleaseContext{
+		cluster:   cluster,
+		manifests: mockManifests,
+		repo:      checkout,
+		registry:  mockRegistry,
+	}
+	for _, tst := range []struct {
+		Name     string
+		Spec     []update.ContainerUpdate
+		Expected update.ControllerResult
+		Commit   string
+	}{
+		{
+			Name: "multiple containers",
+			Spec: []update.ContainerUpdate{
+				{
+					Container: helloContainer,
+					Current:   oldRef,
+					Target:    newHwRef,
+				},
+				{
+					Container: sidecarContainer,
+					Current:   sidecarRef,
+					Target:    newSidecarRef,
+				},
+			},
+			Expected: update.ControllerResult{
+				Status: update.ReleaseStatusSuccess,
+				PerContainer: []update.ContainerUpdate{{
+					Container: helloContainer,
+					Current:   oldRef,
+					Target:    newHwRef,
+				}, {
+					Container: sidecarContainer,
+					Current:   sidecarRef,
+					Target:    newSidecarRef,
+				}},
+			},
+			Commit: "Release containers\n\ndefault:deployment/helloworld\n- quay.io/weaveworks/helloworld:master-a000002\n- weaveworks/sidecar:master-a000002\n",
+		},
+		{
+			Name: "container tag mismatch",
+			Spec: []update.ContainerUpdate{
+				{
+					Container: helloContainer,
+					Current:   newHwRef, // mismatch
+					Target:    oldRef,
+				},
+				{
+					Container: sidecarContainer,
+					Current:   sidecarRef,
+					Target:    newSidecarRef,
+				},
+			},
+			Expected: update.ControllerResult{
+				Status: update.ReleaseStatusFailed,
+				Error:  fmt.Sprintf(update.ContainerTagMismatch, helloContainer),
+			},
+			Commit: "Release containers\n\ndefault:deployment/helloworld failed: container tag mismatch: greeter",
```

Comment:
The success was initially not mentioned because any failure for a container also meant the workload fails. I changed this now to allow partial workload updates. In this case, we don't have access to the errors since the workload can only have one result (success or failed).